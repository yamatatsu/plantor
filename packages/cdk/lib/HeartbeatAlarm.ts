import { App, Stack, StackProps, aws_iotevents, aws_iam } from "aws-cdk-lib";

type Props = StackProps & {};

export class HeartbeatAlarm extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    enum State {
      Normal = "Normal",
      Error = "Error",
    }

    const timer: aws_iotevents.CfnDetectorModel.SetTimerProperty = {
      timerName: "HeartbeatTimeoutTimer",
      seconds: 60,
    };

    const deviceNameJsonPath = "deviceName";
    const topicInput = new aws_iotevents.CfnInput(this, "TopicInput", {
      inputName: "plantor-topic",
      inputDefinition: {
        attributes: [{ jsonPath: deviceNameJsonPath }],
      },
    });

    const normalCondition = `!isUndefined($input.${topicInput.inputName}.${deviceNameJsonPath})`;

    const normalState = {
      stateName: State.Normal,
      onEnter: {
        events: [
          {
            eventName: "createTimer",
            condition: "true",
            actions: [{ setTimer: timer }],
          },
        ],
      },
      onInput: {
        events: [
          {
            eventName: "inputEvent",
            condition: normalCondition,
            actions: [{ resetTimer: timer }],
          },
        ],
        transitionEvents: [
          {
            eventName: "timerTimeout",
            condition: `timeout("${timer.timerName}")`,
            actions: [{ clearTimer: timer }],
            nextState: State.Error,
          },
        ],
      },
    };
    const errorState = {
      stateName: State.Error,
      onEnter: {
        events: [
          {
            eventName: "testPublish",
            condition: "true",
            actions: [{ iotTopicPublish: { mqttTopic: "testFromIotEvent" } }],
          },
        ],
      },
      onInput: {
        transitionEvents: [
          {
            eventName: "inputArrival",
            condition: normalCondition,
            nextState: State.Normal,
          },
        ],
      },
    };

    new aws_iotevents.CfnDetectorModel(this, "HeartbeatAlarmModel", {
      key: deviceNameJsonPath,
      evaluationMethod: "BATCH",
      roleArn: "", // TODO:
      detectorModelDefinition: {
        initialStateName: State.Normal,
        states: [normalState, errorState],
      },
    });
  }
}
