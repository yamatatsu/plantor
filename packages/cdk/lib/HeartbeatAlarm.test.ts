import { SynthUtils } from "@aws-cdk/assert";
import { App } from "aws-cdk-lib";
import { HeartbeatAlarm } from "./HeartbeatAlarm";

test("snapshot test", () => {
  const app = new App();

  const target = new HeartbeatAlarm(app, "Target", {});

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
