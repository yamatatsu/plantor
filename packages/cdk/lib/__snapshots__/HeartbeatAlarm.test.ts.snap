// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "HeartbeatAlarmModel": Object {
      "Properties": Object {
        "DetectorModelDefinition": Object {
          "InitialStateName": "Normal",
          "States": Array [
            Object {
              "OnEnter": Object {
                "Events": Array [
                  Object {
                    "Actions": Array [
                      Object {
                        "SetTimer": Object {
                          "Seconds": 60,
                          "TimerName": "HeartbeatTimeoutTimer",
                        },
                      },
                    ],
                    "Condition": "true",
                    "EventName": "createTimer",
                  },
                ],
              },
              "OnInput": Object {
                "Events": Array [
                  Object {
                    "Actions": Array [
                      Object {
                        "ResetTimer": Object {
                          "TimerName": "HeartbeatTimeoutTimer",
                        },
                      },
                    ],
                    "Condition": "!isUndefined($input.plantor-topic.deviceName)",
                    "EventName": "inputEvent",
                  },
                ],
                "TransitionEvents": Array [
                  Object {
                    "Actions": Array [
                      Object {
                        "ClearTimer": Object {
                          "TimerName": "HeartbeatTimeoutTimer",
                        },
                      },
                    ],
                    "Condition": "timeout(\\"HeartbeatTimeoutTimer\\")",
                    "EventName": "timerTimeout",
                    "NextState": "Error",
                  },
                ],
              },
              "StateName": "Normal",
            },
            Object {
              "OnEnter": Object {
                "Events": Array [
                  Object {
                    "Actions": Array [
                      Object {
                        "IotTopicPublish": Object {
                          "MqttTopic": "testFromIotEvent",
                        },
                      },
                    ],
                    "Condition": "true",
                    "EventName": "testPublish",
                  },
                ],
              },
              "OnInput": Object {
                "TransitionEvents": Array [
                  Object {
                    "Condition": "!isUndefined($input.plantor-topic.deviceName)",
                    "EventName": "inputArrival",
                    "NextState": "Normal",
                  },
                ],
              },
              "StateName": "Error",
            },
          ],
        },
        "EvaluationMethod": "BATCH",
        "Key": "deviceName",
        "RoleArn": "",
      },
      "Type": "AWS::IoTEvents::DetectorModel",
    },
    "TopicInput": Object {
      "Properties": Object {
        "InputDefinition": Object {
          "Attributes": Array [
            Object {
              "JsonPath": "deviceName",
            },
          ],
        },
        "InputName": "plantor-topic",
      },
      "Type": "AWS::IoTEvents::Input",
    },
  },
}
`;
